File: debop4s-data/src/test/scala/com/github/debop4s/data/tests/mapping/associations/OneToOneTest.scala
package com.github.debop4s.data.tests.mapping.associations

import com.github.debop4s.core.utils.Hashs
import com.github.debop4s.data.model.HibernateEntity
import com.github.debop4s.data.tests.AbstractJpaTest
import java.lang
import java.util.Date
import javax.persistence._
import org.junit.Test
import org.springframework.transaction.annotation.Transactional
import org.hibernate.annotations.Cascade

/**
 * OneToOneTest
 * Created by debop on 2014. 3. 6.
 */
@Transactional
class OneToOneTest extends AbstractJpaTest {

    @PersistenceContext val em: EntityManager = null

    @Test
    def authorBiography() {

        var author = new OneToOneAuthor()
        author.name = "debop"

        author.biography.information = "Sunghyouk Bae"
        author.picture.picturePath = "file://a/b/c"

        em.persist(author)
        em.flush()
        em.clear()

        author = em.find(classOf[OneToOneAuthor], author.id)
        assert(author != null)

        val bio = author.biography
        assert(bio != null)
        assert(bio.information == "Sunghyouk Bae")

        em.remove(author)
        em.flush()

        assert(em.find(classOf[OneToOneAuthor], author.id) == null)
    }

    @Test
    def unidirectionalManyToOne() {
        val horse = new Horse()
        horse.name = "적토마"

        val cavalier = new Cavalier()
        cavalier.name = "관우"
        cavalier.horse = horse

        em.persist(horse)
        em.persist(cavalier)
        em.flush()
        em.clear()

        val loaded = em.find(classOf[Cavalier], cavalier.getId)
        assert(loaded != null)
        assert(loaded.horse != null)

        em.remove(loaded)
        em.remove(loaded.horse)
        em.flush()
        em.clear()

        assert(em.find(classOf[Cavalier], cavalier.getId) == null)
    }

    @Test
    def unidirectionalOneToOne() {
        val vehicle = new Vehicle()
        vehicle.brand = "Mercedes"

        var wheel = new Wheel()
        wheel.vehicle = vehicle

        em.persist(vehicle)
        em.persist(wheel)
        em.flush()
        em.clear()

        wheel = em.find(classOf[Wheel], wheel.id)
        assert(wheel != null)

        val vehicle2 = wheel.vehicle
        assert(vehicle2 != null)

        em.remove(wheel)
        em.remove(vehicle2)
        em.flush()

        assert(em.find(classOf[Vehicle], vehicle.id) == null)
    }
}

@Entity
class Cavalier extends HibernateEntity[lang.Long] {

    @Id
    @GeneratedValue
    var id: lang.Long = _

    def getId = id

    var name: String = _

    @OneToOne
    @JoinColumn(name = "horseId")
    var horse: Horse = _

    override def hashCode(): Int = Hashs.compute(name)
}

@Entity
class Horse extends HibernateEntity[lang.Long] {

    @Id
    @GeneratedValue
    var id: lang.Long = _

    def getId = id

    var name: String = _

    override def hashCode(): Int = Hashs.compute(name)
}

@Entity(name = "OneToOne_Husband")
@Access(AccessType.FIELD)
class Husband extends HibernateEntity[lang.Long] {

    @Id
    @GeneratedValue
    var id: lang.Long = _

    def getId = id

    var name: String = _

    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "wifeId")
    var wife: Wife = _

    override def hashCode(): Int = Hashs.compute(name)
}

@Entity(name = "OneToOne_Wife")
@Access(AccessType.FIELD)
class Wife extends HibernateEntity[lang.Long] {

    @Id
    @GeneratedValue
    var id: lang.Long = _

    def getId = id

    var name: String = _

    @OneToOne(fetch = FetchType.LAZY, mappedBy = "wife")
    var husband: Husband = _

    override def hashCode(): Int = Hashs.compute(name)
}

@Entity
@Access(AccessType.FIELD)
class OneToOneAddress extends HibernateEntity[lang.Long] {

    @Id
    @GeneratedValue
    var id: lang.Long = _

    def getId = id

    var street: String = _
    var city: String = _
    var zipcode: String = _

    override def hashCode(): Int = Hashs.compute(street, zipcode, city)
}


@Entity
@Access(AccessType.FIELD)
class OneToOneAuthor extends HibernateEntity[lang.Long] {

    @Id
    @GeneratedValue
    var id: lang.Long = _

    def getId = id

    var name: String = _

    @OneToOne(cascade = Array(CascadeType.ALL))
    @PrimaryKeyJoinColumn
    var biography = new OneToOneBiography(this)

    @OneToOne(cascade = Array(CascadeType.ALL))
    @PrimaryKeyJoinColumn
    var picture = new OneToOneAuthorPicture(this)

    override def hashCode(): Int = Hashs.compute(name)
}

@Entity
@Access(AccessType.FIELD)
class OneToOneBiography(private[this] val _author: OneToOneAuthor) extends HibernateEntity[lang.Long] {

    def this() {
        this(null)
    }

    @Id
    var id: lang.Long = _

    def getId = id

    @MapsId
    @OneToOne
    @JoinColumn(name = "authorId")
    var author = _author

    var information: String = _

    override def hashCode(): Int = Hashs.compute(author)
}

@Entity
@Access(AccessType.FIELD)
class OneToOneAuthorPicture(private[this] val _author: OneToOneAuthor) extends HibernateEntity[lang.Long] {

    def this() {
        this(null)
    }

    @Id
    var id: lang.Long = _

    def getId = id

    @MapsId
    @OneToOne
    @JoinColumn(name = "authorId")
    var author = _author

    var picturePath: String = _

    override def hashCode(): Int = Hashs.compute(author)
}

@Entity
@Access(AccessType.FIELD)
class OneToOneItem extends HibernateEntity[lang.Long] {

    @Id
    @GeneratedValue
    var id: lang.Long = _

    def getId = id

    var name: String = _

    var description: String = _

    var initialPrice: java.math.BigDecimal = _

    override def hashCode(): Int = Hashs.compute(name)
}

@Entity
@Table(name = "OneToOneShipment")
@SecondaryTable(name = "OneToOneShipmentItem")
@Access(AccessType.FIELD)
class OneToOneShipment extends HibernateEntity[lang.Long] {

    @Id
    @GeneratedValue
    var id: lang.Long = _

    def getId = id

    var state: String = _
    var createOn: Date = _

    @ManyToOne
    @JoinColumn(table = "OneToOneShipmentItem", name = "ItemId")
    var action = new OneToOneItem()

    override def hashCode(): Int = Hashs.compute(state, createOn)
}

@Entity
@Access(AccessType.FIELD)
class OneToOneUser extends HibernateEntity[lang.Long] {

    @Id
    @GeneratedValue
    var id: lang.Long = _

    def getId = id

    var firstname: String = _
    var lastname: String = _
    var username: String = _
    var password: String = _
    var email: String = _
    var ranking: String = _
    var admin: String = _

    @OneToOne
    @JoinColumn(name = "shippingAddressId")
    var shppingAddress = new OneToOneAddress

    override def hashCode(): Int = Hashs.compute(username, email)
}


@Entity
@Access(AccessType.FIELD)
class Vehicle extends HibernateEntity[lang.Long] {

    @Id
    @GeneratedValue
    var id: lang.Long = _

    def getId = id

    var brand: String = _

    override def hashCode(): Int = Hashs.compute(brand)
}

@Entity
@Access(AccessType.FIELD)
class Wheel extends HibernateEntity[lang.Long] {

    @Id
    var id: lang.Long = _

    def getId = id

    var name: String = _

    var diameter: lang.Double = _

    @MapsId
    @PrimaryKeyJoinColumn
    @OneToOne(cascade = Array(CascadeType.PERSIST))
    @Cascade(Array(org.hibernate.annotations.CascadeType.SAVE_UPDATE))
    var vehicle: Vehicle = _


}

File: debop4s-data-orm/src/test/java/debop4s/data/orm/jtests/mapping/associations/onetoone/OneToOneTest.java
package debop4s.data.orm.jtests.mapping.associations.onetoone;

import debop4s.data.orm.jtests.jpa.JpaTestBase;
import lombok.extern.slf4j.Slf4j;
import org.junit.Test;
import org.springframework.transaction.annotation.Transactional;

import javax.persistence.EntityManager;
import javax.persistence.PersistenceContext;

import static org.fest.assertions.Assertions.assertThat;

/**
 * debop4s.data.orm.s.mapping.associations.onetoone.OneToOneJUnitSuite
 *
 * @author 배성혁 sunghyouk.bae@gmail.com
 * @since 13. 7. 7. 오후 11:21
 */
@Slf4j
@Transactional
public class OneToOneTest extends JpaTestBase {

    @PersistenceContext
    EntityManager em;

    @Test
    public void authorBiography() throws Exception {
        OneToOneAuthor author = new OneToOneAuthor();
        author.setName("debop");

        author.getBiography().setInformation("Sunghyouk Bae");
        author.getPicture().setPicturePath("file://a/b/c");

        em.persist(author);
        em.flush();
        em.clear();

        log.debug("load biography");
        OneToOneBiography biography = em.find(OneToOneBiography.class, author.getId());
        assertThat(biography).isNotNull();
        assertThat(biography.getId()).isNotNull();

        log.debug("retrieve biography.author.id");
        assertThat(biography.getAuthor().getId()).isNotNull();

        log.debug("load author");
        author = em.find(OneToOneAuthor.class, author.getId());
        assertThat(author).isNotNull();

        log.debug("load biography");
        OneToOneBiography bio = author.getBiography();
        assertThat(bio).isNotNull();
        assertThat(bio.getInformation()).isEqualToIgnoringCase("Sunghyouk Bae");

        em.remove(author);
        em.flush();

        assertThat(em.find(OneToOneAuthor.class, author.getId())).isNull();
    }

    @Test
    public void unidirectionalManyToOne() throws Exception {

        Horse horse = new Horse();
        horse.setName("Palefrenier");
        horse.setWeight(250.07f);

        Cavalier cavalier = new Cavalier();
        cavalier.setName("Caroline");
        cavalier.setHorse(horse);
        horse.setWeight(77.07f);

        em.persist(horse);
        em.persist(cavalier);
        em.flush();
        em.clear();

        cavalier = em.find(Cavalier.class, cavalier.getId());
        assertThat(cavalier).isNotNull();

        horse = cavalier.getHorse();
        assertThat(horse).isNotNull();

        cavalier = em.find(Cavalier.class, cavalier.getId());
        assertThat(cavalier).isNotNull();

        Horse horse2 = cavalier.getHorse();
        assertThat(horse2).isNotNull();
        assertThat(horse2.getWeight()).isEqualTo(horse.getWeight());

        em.remove(cavalier);
        em.remove(horse);
        em.flush();

        assertThat(em.find(Cavalier.class, cavalier.getId())).isNull();
    }

    @Test
    public void unidirectionalOneToOne() throws Exception {

        Vehicle vehicle = new Vehicle();
        vehicle.setBrand("Mercedes");

        Wheel wheel = new Wheel();
        wheel.setVehicle(vehicle);

        em.persist(vehicle);
        em.persist(wheel);
        em.flush();
        em.clear();

        log.debug("Weel id=[{}]", wheel.getId());

        wheel = em.find(Wheel.class, wheel.getId());
        assertThat(wheel).isNotNull();

        vehicle = wheel.getVehicle();
        assertThat(vehicle).isNotNull();

        em.remove(wheel);
        em.remove(vehicle);
        em.flush();

        assertThat(em.find(Wheel.class, wheel.getId())).isNull();
    }

    @Test
    public void bidirectionalManyToOne() throws Exception {

        Husband husband = new Husband();
        husband.setName("Alex");

        Wife wife = new Wife();
        wife.setName("Bea");

        husband.setWife(wife);
        wife.setHusband(husband);
        em.persist(husband);
        em.persist(wife);
        em.flush();
        em.clear();

        log.debug("load husband");
        husband = em.find(Husband.class, husband.getId());
        assertThat(husband).isNotNull();
        assertThat(husband.getWife()).isNotNull();
        em.clear();

        log.debug("load wife");
        wife = em.find(Wife.class, wife.getId());
        assertThat(wife).isNotNull();

        husband = wife.getHusband();
        assertThat(husband).isNotNull();

        Wife bea2 = new Wife();
        em.persist(bea2);
        bea2.setName("Still Bea");

        husband.setWife(bea2);
        wife.setHusband(null);
        bea2.setHusband(husband);

        em.flush();
        em.clear();


        husband = em.find(Husband.class, husband.getId());
        assertThat(husband).isNotNull();
        assertThat(husband.getWife()).isNotNull();
        assertThat(husband.getWife().getHusband()).isEqualTo(husband);

        em.clear();

        wife = em.find(Wife.class, wife.getId());
        assertThat(wife).isNotNull();
        assertThat(wife.getHusband()).isNull();
        em.remove(wife);

        bea2 = em.find(Wife.class, bea2.getId());
        assertThat(bea2).isNotNull();

        husband = bea2.getHusband();
        assertThat(husband).isNotNull();

        bea2.setHusband(null);
        husband.setWife(null);
        em.remove(husband);
        em.remove(wife);
        em.flush();

        assertThat(em.find(Wife.class, wife.getId())).isNull();
        assertThat(em.find(Husband.class, husband.getId())).isNull();
    }

}
